File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int solution(int[] numbers, int target) 
    {
        int answer = 0;
        //합계와 인덱스를 담음
        Queue<(int sum, int index)> ans = new Queue<(int sum, int index)>();
        
        ans.Enqueue((0, 0));
        while(ans.Count > 0)
        {
            //현재 합과 위치 저장
            var current = ans.Dequeue();
            int currentSum = current.sum;
            int currentIndex = current.index;
            
            //인덱스의 마지막
            if(currentIndex == numbers.Length)
            {
                if(currentSum == target) answer++;    
            }
            else
            {
                //더한경우
                ans.Enqueue((currentSum + numbers[currentIndex], currentIndex+1));   
                //뺀 경우
                ans.Enqueue((currentSum - numbers[currentIndex], currentIndex+1)); 
            }  
        }
        return answer;
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.io.*;
import java.util.*;

class Solution {
    public int solution(int[] numbers, int target) 
    {
        int answer = 0;
        
        answer = DFS(numbers, target, 0, 0);
        return answer;
    }
    
    public int DFS(int[] numbers, int target, int index, int total)
    {
        //배열의 마지막에 도달
        if(index == numbers.length)
        {
            //최종값이 target일 경우
            if(total == target)
            {
                return 1;  
            } 
            else
            {
                return 0;    
            }
        }
        
        int total_add = DFS(numbers, target, index+1, total + numbers[index]); //더하는 경우
        int total_abs = DFS(numbers, target, index+1, total - numbers[index]); //빼는 경우
        
        return total_add + total_abs;
    }
}